File: src/game/state.ts
export type Player = number;
export type Owner = Player | null;
export type NodeId = number;

export type Vec2 = { x: number; y: number };

export type GameNode = {
  id: NodeId;
  pos: Vec2;
  owner: Owner;
  strength: number;
};

export type Edge = { a: NodeId; b: NodeId };

export type Flow = { src: NodeId; dst: NodeId; player: Player };

export type GameState = {
  nodes: GameNode[];
  edges: Edge[];
  flows: Flow[];
  tick: number;
  numPlayers: number;
};

export type Action =
  | { kind: 'toggleFlow'; src: NodeId; dst: NodeId; player: Player };

export const REGEN_PER_SEC = 1.0;
export const TRANSFER_PER_SEC = 3.0;
export const MAX_STRENGTH = 100;
export const MIN_STRENGTH_TO_SEND = 0.1;
export const LOOP_BONUS = 0.5;

File: python/flux/state.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional, List, Literal

Player = int
Owner = Optional[int]
NodeId = int


@dataclass(frozen=True)
class Vec2:
    x: float
    y: float


@dataclass(frozen=True)
class GameNode:
    id: NodeId
    pos: Vec2
    owner: Owner
    strength: float


@dataclass(frozen=True)
class Edge:
    a: NodeId
    b: NodeId


@dataclass(frozen=True)
class Flow:
    src: NodeId
    dst: NodeId
    player: Player


@dataclass(frozen=True)
class GameState:
    nodes: tuple[GameNode, ...]
    edges: tuple[Edge, ...]
    flows: tuple[Flow, ...]
    tick: int
    num_players: int


@dataclass(frozen=True)
class Action:
    kind: Literal["toggleFlow"]
    src: NodeId
    dst: NodeId
    player: Player


REGEN_PER_SEC = 1.1
TRANSFER_PER_SEC = 3.0
MAX_STRENGTH = 100.0
MIN_STRENGTH_TO_SEND = 0.1
ATTACK_BONUS = 0.5
